aptos-move/block-executor/src/executor.rs

```diff
@@ -216,16 +226,57 @@ where
 
         let _timer = WORK_WITH_TASK_SECONDS.start_timer();
         let mut scheduler_task = SchedulerTask::NoTask;
+        let mut accumulated_gas = 0;
         loop {
             // Only one thread try_commit to avoid contention.
             if committing {
                 // Keep committing txns until there is no more that can be committed now.
                 while let Some(txn_idx) = scheduler.try_commit() {
+                    // Committed the last transaction, BlockSTM finishes execution.
                     if txn_idx as usize + 1 == block.len() {
-                        // Committed the last transaction / everything.
                         scheduler_task = SchedulerTask::Done;
+
+                        counters::PARALLEL_PER_BLOCK_GAS.observe(accumulated_gas as f64);
                         break;
                     }
+
+                    // For committed txns with Success status, calculate the accumulated gas.
+                    // For committed txns with Abort or SkipRest status, early halt BlockSTM.
+                    match last_input_output.success_gas(txn_idx) {
+                        Ok(gas) => {
+                            accumulated_gas += gas;
+                            counters::PARALLEL_PER_TXN_GAS.observe(gas as f64);
+                        },
+                        _ => {
+                            scheduler.halt();
```

Comment:
I think even for aptos-vm speculative logging, currently it wouldn't work because speculative logs are flushed at the end of the block execution (in aptos-vm/block-executor/mod.rs), so the errors that are speculatively generated by halt for the skipped transactions would still be logged.

Eventually we could flush on commit_hook, but we can't do that now because module r/w fallback can happen anytime (and it correctly clears everything by re-init). I see two ways to resolve this:
(1) in flush providing a number (of successful / committed transactions), and ignoring the rest.
(2) clearing all the logs for skipped transactions manually - this seems a lot trickier and uglier.

flushing logic is in aptos-logger/src/lib.rs. May want to move the flushing from mod.rs to execute block itself as it may be easier to propagate the number of transactions to flush for, but you should decide what's best.